echo-telemetry/src/main/java/com/netflix/spinnaker/echo/telemetry/TelemetryEventListener.java

```diff
@@ -0,0 +1,170 @@
+import com.netflix.spinnaker.kork.proto.stats.Execution;
+import com.netflix.spinnaker.kork.proto.stats.SpinnakerInstance;
+import com.netflix.spinnaker.kork.proto.stats.Stage;
+import com.netflix.spinnaker.kork.proto.stats.Status;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+import retrofit.mime.TypedString;
+
+@Slf4j
+@Component
+@ConditionalOnProperty("telemetry.enabled")
+public class TelemetryEventListener implements EchoEventListener {
+
+  private static final Set<String> LOGGABLE_DETAIL_TYPES =
+      ImmutableSet.of(
+          "orca:orchestration:complete",
+          "orca:orchestration:failed",
+          "orca:pipeline:complete",
+          "orca:pipeline:failed");
+
+  private static final JsonFormat.Printer JSON_PRINTER =
+      JsonFormat.printer().omittingInsignificantWhitespace();
+
+  private final TelemetryService telemetryService;
+
+  private final TelemetryConfig.TelemetryConfigProps telemetryConfigProps;
+
+  @Autowired
+  public TelemetryEventListener(
+      TelemetryService telemetryService,
+      TelemetryConfig.TelemetryConfigProps telemetryConfigProps) {
+    this.telemetryService = telemetryService;
+    this.telemetryConfigProps = telemetryConfigProps;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public void processEvent(Event event) {
+    try {
+      if (event.getDetails() == null || event.getContent() == null) {
+        log.debug("Telemetry not sent: Details or content not found in event");
+        return;
+      }
+
+      String eventType = event.getDetails().getType();
+      if (!LOGGABLE_DETAIL_TYPES.contains(eventType)) {
+        log.debug("Telemetry not sent: type '{}' not whitelisted ", eventType);
+        return;
+      }
+
+      String applicationId = event.getDetails().getApplication();
+      if (applicationId == null || applicationId.isEmpty()) {
+        log.debug("Application ID must be non-null and not empty");
+        return;
+      }
+
+      Map<String, Object> execution = (Map<String, Object>) event.getContent().get("execution");
+      if (execution == null || execution.isEmpty()) {
+        log.debug("Missing execution from Event content.");
+        return;
+      }
+
+      String hashedApplicationId = hash(applicationId);
+      Execution.Type executionType =
+          Execution.Type.valueOf(
+              // TODO(ttomsu, louisjimenez): Add MPTv1 and v2 execution type detection.
+              execution.getOrDefault("type", "UNKNOWN").toString().toUpperCase());
+      Status executionStatus =
+          Status.valueOf(execution.getOrDefault("status", "UNKNOWN").toString().toUpperCase());
+
+      Map trigger = (Map) execution.getOrDefault("trigger", new HashMap());
```

Comment:
Maybe make this a `Map<String, Object>` to provide some type safety given that we're assuming that the keys are all strings.  (We're actually also assuming all values are strings as well, I believe, so you could probably also do `Map<String, String>`.)

Reply:
We're assuming all keys are strings, but not all of the values are strings.

Added (String,Object) for type safety